### PR TITLE
Fix firmwares loading for iMX6

### PR DESCRIPTION
**Pull request by by-gnome**
- include SDMA and VPU firmware from firmware-imx into kernel 
- configure Coda codec IP as built-in
- configure brcmfmac driver and cfg80211 as modules
- move from brcmfmac_sdio-firmware-imx to brcmfmac_sdio-firmware

Because i.MX SDMA support is built-in so need to make the firmware built-in.
Since SDMA and VPU firmware are in the same package make Coda codec IP built-in too.
To allow brcmfmac driver and cfg80211 can load firmware from the filesystem make them as modules.
Package brcmfmac_sdio-firmware-imx is obsolete, move to brcmfmac_sdio-firmware.

### CODE REVIEW

**Comment by jernejsk** [comment]
why does brcmfmac driver needs build in firmware? all other projects have it separately and it works just fine

**Comment by by-gnome** [comment]
Because brcmfmac driver and cfg80211 are built-in. This is how it was configured.

**Comment by lrusak** [comment]
> Because brcmfmac driver and cfg80211 are built-in. This is how it was configured.

Can we change it to `CONFIG_BRCMFMAC=m` and `CONFIG_CFG80211=m` then?

**Comment by jernejsk** [comment]
yes, you said that in PR message, but I don't see a connection - firmware loading has independent mechanism to look for a file on file system - why wouldn't that work? and if that really doesn't work, preferred method would be do change brcmfmac to module.

**Comment by by-gnome** [comment]
@lrusak No problem, i will reconfigure as modules.
@jernejsk because kernel-overlays with firmware becomes available some later, after brcmfmac driver and cfg802111 try to load firmware.

**Comment by chewitt** [comment]
In other SoC projects it makes no difference whether the drivers are `=y` or `=m` the hardware still probes and the firmware still loads from the overlay dirs and none of the firmware is embedded into the kernel itself. If there is a timing issue, please show evidence of a problem via boot logs (dmesg) so we can see; the solution will be to get kernel devs to fix the timing isssue, not to bulid firmware in. 

Please also check whether the broadcom firmwares are available in the kernel-firmware package, as i.MX6 is normally well supported upstream and we should prefer the kernel firmware repo (linux-firmware in LE) over our own local firmware hoard with a long-term view on dropping the local package.

**Comment by by-gnome** [comment]
@chewitt first log - brcmfmac and cfg80211 as modules: [dmesg_12_wb_c1_brcmfmac_module.txt](https://github.com/LibreELEC/LibreELEC.tv/files/5580571/dmesg_12_wb_c1_brcmfmac_module.txt)
second log - brcmfmac and cfg80211 built-in: [dmesg_12_wb_c1_brcmfmac_builtin.txt](https://github.com/LibreELEC/LibreELEC.tv/files/5580572/dmesg_12_wb_c1_brcmfmac_builtin.txt)

In the kernel-firmware we can find binary for wifi: brcmfmacXXXX-sdio.bin, but we need for wifi brcmfmacXXXX-sdio.txt also, and BCMXXXX.hcd for bluetooth. So it may be comdbination of kernel-firmware and brcmfmac_sdio-firmware (i can skip binaries using brcmfmac_sdio-firmware.dat)

**Comment by chewitt** [comment]
The logs both show the broadcom firmwares loaded fine. I'd guess the boot is a little faster with `=y` so that's fine to keep, and it's the same in several other projects. Thanks for checking the kernel-firmware status; I'm going to drop Josh Boyer a note to see if/how he'd like us to upstream a load of broadcom firmware and nvram files. 

So last on the list is the `imx-firmware` package. Why the change to build firmware into the kernel? It doesn't look like it was needed before, and AFAIK we don't do it for any other ARM projects.

**Comment by by-gnome** [comment]
SDMA the same reason - firmware didn't loaded.
This log I made from master before my first PR: [dmesg_1_wb_c1_master.txt](https://github.com/LibreELEC/LibreELEC.tv/files/5581554/dmesg_1_wb_c1_master.txt) 
all firmware are in filesystem - no firmware were loaded except VPU (coda):
```
[    2.064703] imx-sdma 20ec000.sdma: Direct firmware load for imx/sdma/sdma-imx6q.bin failed with error -2
[    2.064862] imx-sdma 20ec000.sdma: external firmware not found, using ROM firmware
[    2.663529] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[    2.664034] brcmfmac mmc1:0001:1: Direct firmware load for brcm/brcmfmac4330-sdio.bin failed with error -2
[    4.035452] platform regulatory.0: Direct firmware load for regulatory.db failed with error -2
[    4.035477] cfg80211: failed to load regulatory.db
```

**Comment by by-gnome** [comment]
> The logs both show the broadcom firmwares loaded fine.

I see in _dmesg_12_wb_c1_brcmfmac_builtin.txt_:
```
[    3.614404] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[    3.615436] brcmfmac mmc1:0001:1: Direct firmware load for brcm/brcmfmac4330-sdio.bin failed with error -2
[    5.295601] platform regulatory.0: Direct firmware load for regulatory.db failed with error -2
[    5.295628] cfg80211: failed to load regulatory.db
```

**Comment by by-gnome** [comment]
@jernejsk @chewitt I just check other projects and... all projects which use brcmfmac_sdio-firmware have CONFIG_CFG80211=m and CONFIG_MAC80211=m in linux config.
Still want to talk about it?

**Comment by lrusak** [comment]
> @jernejsk @chewitt I just check other projects and... all projects which use brcmfmac_sdio-firmware (Allwinner, Amlogic, RPi, Rockchip) have CONFIG_CFG80211=m and CONFIG_MAC80211=m in linux config.
> Still want to talk about it?

Yes it would be nice IMO if you could use `CONFIG_CFG80211=m` and `CONFIG_MAC80211=m` as to be synchronised with the other projects

File: packages/linux/package.mk [hunk 1]
```diff
@@ -45,6 +45,8 @@ fi
 if [ "$TARGET_ARCH" = "x86_64" ]; then
   PKG_DEPENDS_TARGET="$PKG_DEPENDS_TARGET elfutils:host pciutils"
   PKG_DEPENDS_UNPACK+=" intel-ucode kernel-firmware"
+elif [ "$TARGET_ARCH" = "arm" -a "$DEVICE" = "iMX6" ]; then
+  PKG_DEPENDS_UNPACK="$PKG_DEPENDS_UNPACK firmware-imx"
```
Comment (chewitt) [inline packages/linux/package.mk:49]: `PKG_DEPENDS_UNPACK+=" firmware-imx"`

**Comment by by-gnome** [comment]
I made log with CONFIG_IMX_SDMA=m: [dmesg_3_wb_c1_sdma_module.txt](https://github.com/LibreELEC/LibreELEC.tv/files/5585384/dmesg_3_wb_c1_sdma_module.txt)
Audio devices imx6-wandboard-sgtl5000 and imx-spdif are broken, SDMA firmware from filesystem has been loaded.

There are two ways:
1. make SDMA firmware built-in from kernel-firmware or firmware-imx.
2. configure Coda codec IP as built-in and make SDMA & VPU firmware both built-in from firmware-imx (currently in PR).

File: packages/linux/package.mk (hunk 1 above)
Comment (by-gnome) [reply packages/linux/package.mk]: Done.

**Comment by by-gnome** [comment]
One more point related to the transition to brcmfmac_sdio-firmware package.
To download the patchram, need to use _udev_ rule and _systemd_ service to run _hciattach_ (actualy two services - one for CuBox-i and the second for Wandboard, because they use different _tty_) or use homemade patches for CuBox-i and Wandboard device trees to add apropriate bluetooth node.
I can add in this PR for completeness.

**Comment by chewitt** [comment]
Since the move to serdev (some time ago now) there is really no need to use hciattach anymore, so the preferred approach is to patch the board dts file(s) and then upstream patches to the kernel so we can drop them with a future kernel bump.

If you've never submitted patches to the kernel before it's not _that_ hard, there's just some etiquette to know/obey (or you learn it the hard way via trial/error) but we can guide that process if needed.

**Comment by by-gnome** [comment]
This is tested on Wandboard B1 and C1.
[dmesg_13_wb_b1.txt](https://github.com/LibreELEC/LibreELEC.tv/files/5593174/dmesg_13_wb_b1.txt) (not supported in master branch yet)
[dmesg_13_wb_c1.txt](https://github.com/LibreELEC/LibreELEC.tv/files/5593175/dmesg_13_wb_c1.txt)

**Comment by by-gnome** [comment]
`dmesg` log from CuBox-i:
[dmesg_13_cuboxi.txt](https://github.com/LibreELEC/LibreELEC.tv/files/5651930/dmesg_13_cuboxi.txt)

**Review by chewitt — APPROVED** [review]

**Comment by chewitt** [comment]
спасибо всё :)

**Comment by sime** [comment]
If possible could the next build that has these changes be linked here?

Thinking for the sake of keen to test iMX6 users out there.

**Comment by chewitt** [comment]
Next build will be here: https://test.libreelec.tv in the morning. Search on `iMX6` to filter the images available.

**Pull request merged**